### PR TITLE
Adding Winograd v21 fused

### PR DESCRIPTION
**Pull request by Slimakanzer**
This PR adds support for winograd v21_1_2 kernels into fusion. Resolves issue #755.

Also fixes two related bugs:
1. `supported_arch` were not checked at **fusion** compilation. Because of what fusion could launch unsupported kernels.
2. `supported_arch` for _winograd v9_2_7_ relu operations was not specified in the _MDGraph_. For that reason kernel could run on a device that is not supported:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/b7630f76dad1926e224a23d5e470e44987d84fc3/src/md_graph.cpp#L579-L584

## Summary of code changes
* Check device compatibility and kernel when **fusion** Compile
* Added `ConvBinWinogradRxSf2x3g1Fused` solver 
* Added tensor/op parameters required for the _v21_1_2_ winograd version
* Added _winograd v21_1_2_ kernel to the Fusion MDGraph

## Local verification
* gfx1030, rocm 3.10
* gfx906, rocm 4.0
* gfx803, rocm 3.10
* OCL + HIP backend

### CODE REVIEW

**Comment by atamazov** [comment]
This does not resolve #763 because that issue is about a bug in the unit test. I will update PR description.

PS Please avoid using automatic closing of issues.

File: src/fusion.cpp [hunk 1]
```diff
@@ -263,6 +270,22 @@ bool ConvForwardOpDescriptor::GetOpAttr(const std::string& sym, int& val) const
     {
         val = base_desc.GetGroupCount();
     }
+    else if(sym == "f_byte_stride_nk")
+    {
+        val = f_strides[0];
```
Comment (Kirpich30000) [inline src/fusion.cpp:275]: What about bwd/wrw? Will it work correctly?

Comment (Slimakanzer) [reply src/fusion.cpp:275]: Fusion API introduces only forward operations like:
* Convolution Forward
* Activation Forward
* Bias Forward

This is explicitly defined in Winograd constraints:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/8c172d466d25146656ae0a53f03e82ab3cb76d30/src/md_graph.cpp#L521-L524

**Comment by atamazov** [comment]
Please merge from develop (see mail)

**Comment by atamazov** [comment]
@Kirpich30000 Is it ready for merge?

**Comment by atamazov** [comment]
@Slimakanzer Do we still need ConvBinWinogradRxSFused Solver? Related .s files?

File: src/md_graph.cpp [hunk 2]
```diff
@@ -476,215 +476,255 @@ static std::vector<DefaultKernelArg> WinogradNodeArgs()
...
+                FusionMDGraph_Edge_Map map_wino_conv;
+                map_wino_conv["constraints"] = {stride_constr};
+                add_meta_wino(map_wino_conv, 5);
+                g.AddEdge(nullptr, vc, map_wino_conv);
+
+                // add 3x3 with higher priority since its the fastest case
+                FusionMDGraph_Edge_Map map_wino_conv_xe3;
+                map_wino_conv_xe3["constraints"] = {stride_constr, "(y == 3) & (x == 3)"};
+                add_meta_wino(map_wino_conv_xe3, 100);
+                g.AddEdge(nullptr, vc, map_wino_conv_xe3);
```
Comment (Kirpich30000) [inline src/md_graph.cpp:719]: FYI winograd might not be the best for some configurations even if the filter size is 3x3

File: src/ocl/fusionopbiasbnactivocl.cpp [hunk 3]
```diff
@@ -8,7 +8,8 @@ namespace fusion {
 bool IsWinograd(const std::vector<solver::AnySolver>& ss)
 {
     assert(ss.size() == 1);
-    return ss[0].GetSolverDbId() == "ConvBinWinogradRxSFused";
+    auto solverId = ss[0].GetSolverDbId();
+    return (solverId == "ConvBinWinogradRxSFused" || solverId == "ConvBinWinogradRxSf2x3g1Fused");
```
Comment (Kirpich30000) [inline src/ocl/fusionopbiasbnactivocl.cpp:12]: Relying on kernel name looks too hacky. @atamazov what is your opinion?

File: src/solver/conv_winoRxS_f2x3.cpp [hunk 4]
```diff
@@ -867,5 +867,42 @@ ConvSolution ConvBinWinogradRxSf2x3g1::GetSolution(const ConvolutionContext& par
     return tunable.GetSolution(params, tunable.GetPerformanceConfig(params), false);
 }
 
+bool ConvBinWinogradRxSf2x3g1Fused::IsApplicable(const ConvolutionContext&) const
+{
+    return true; // Actual checks moved to FusionMDGraph.
```
Comment (Kirpich30000) [inline src/solver/conv_winoRxS_f2x3.cpp:872]: This one looks suspicious. Might it break some usecases?

File: src/ocl/fusionopbiasbnactivocl.cpp (hunk 3 above)
Comment (atamazov) [reply src/ocl/fusionopbiasbnactivocl.cpp:12]: Inevitable for now.

File: src/md_graph.cpp (hunk 2 above)
Comment (atamazov) [reply src/md_graph.cpp:719]: You mean v21 Winograd could be slower than older fused Winograd?

File: src/solver/conv_winoRxS_f2x3.cpp (hunk 4 above)
Comment (atamazov) [reply src/solver/conv_winoRxS_f2x3.cpp:872]: If implementation in MDGraph is correct, then no.

**Comment by Slimakanzer** [comment]
> @Slimakanzer Do we still need ConvBinWinogradRxSFused Solver? Related .s files?

@atamazov `ConvBinWinogradRxSFused` provides _gfx8_ compatible kernels, however the previous _gfx9_ winograd kernels are no longer needed (removed in da35c78). Unfortunately, the _v21_ and _v9_2_7_ metadata cannot be combined because they have a different set of arguments.

**Review by Kirpich30000 — APPROVED** [review]
LGTM

**Review by atamazov — APPROVED** [review]
Ok

**Pull request merged**